### ISSUE TITLE
Deleted user still as member of the project space

### ISSUE THREAD

**Issue by ScharfViktor** [owncloud/ocis#4455]
ownCloud Web UI 5.7.0-rc.10 
Infinite Scale 2.0.0-beta.7

Steps:

- admin creates space
- admin adds users "richard" and "einstein" as members
- admin creates public link to the space
- admin deletes user "richard"
- admin opens space

GET `https://localhost:9200/graph/v1.0/me/drives?%24top=0&%24skip=0&%24orderby=name+asc&%24filter=driveType+eq+project`

richard still as member:
![Screenshot 2022-08-24 at 16 57 29](https://user-images.githubusercontent.com/84779829/186452429-339c3958-d310-4d98-8f0f-2bc4ce135b66.png)

- admin opens members

web try to call deleted richard and get 502 GET `https://localhost:9200/graph/v1.0/users/932b4540-8d16-481e-8ef4-588e4b6b151c?%24select=&%24expand=drive`

we haven't on the web no link no user einstein and cannot add new user member: 
![Screenshot 2022-08-24 at 17 03 55](https://user-images.githubusercontent.com/84779829/186453956-46573555-ffd8-46e4-b406-1988adab251e.png)

related https://github.com/owncloud/ocis/issues/903

**Comment by dragonchaser** [comment]
@butonic not sure, IMHO it makes sense that to delete all deleted users from the spaces. Marking users as disabled would make things more complex (ui changes where you can not select disabled users from dropdowns when inviting them, adding visuals to mark disabled users etc......)

**Comment by dragonchaser** [comment]
This will create a bunch of issues, after a lengthy discussion with @butonic and @mmattel  we identified the following:

- in decomposedfs `reva/pkg/storage/utils/decomposedfs/grants.go` the AddGrant call allows the call when neither an owner nor a grant is present:

  ```
	// If the owner is empty and there are no grantees then we are dealing with a just created project space.
	// In this case we don't need to check for permissions and just add the grant since this will be the project
	// manager.
	
  ```

  when we start removing users from spaces (and shares) when they are deleted, it can lead to a space that has no grant or owner .... which would allow anyone to claim ownership or space manager permissions on that space.

- @mmattel pointed out that in addition to a list-all-spaces we need a manage-all-spaces permission that allows (only admins) admins to add users to spaces but does not allow them access to the content of the space.

- Adding grants will be logged in the audit log, so if admins add themself to a space they should not access they can get fired ...

So what would be the best choice to tackle this before GA without opening the full can of worms? :worm: :worm:

**Comment by rhafer** [comment]
> * @mmattel pointed out that in addition to a list-all-spaces we need a manage-all-spaces permission that allows (only admins) admins to add users to spaces but does not allow them access to the content of the space.

This was raised here already #4877.  Which is closed. If there is anything new to discuss, I guess we should reopen it.

**Comment by mmattel** [comment]
I am in favour of reopening the referenced issue as the general topic **granting space rights** vs **content access** was imho not fully considered.

**Comment by micbar** [comment]
@dragonchaser @butonic @rhafer 

IMO this is far out of our GA scope.

For the moment, this limitation is better than opening the can of worms of a "can-manage-access-on-all-spaces" permission. We currently have no UI for that, the webUI only uses `/me/drives` and is not aware of the `/drives/` endpoint.

**Comment by kulmann** [comment]
> @dragonchaser @butonic @rhafer
> 
> IMO this is far out of our GA scope.
> 
> For the moment, this limitation is better than opening the can of worms of a "can-manage-access-on-all-spaces" permission. We currently have no UI for that, the webUI only uses `/me/drives` and is not aware of the `/drives/` endpoint.

I fully agree. Far beyond the scope of our GA.

**Comment by phil-davis** [comment]
> in addition to a list-all-spaces we need a manage-all-spaces permission that allows (only admins) admins to add users to spaces but does not allow them access to the content of the space.
Adding grants will be logged in the audit log, so if admins add themself to a space they should not access they can get fired 

Agree. I will just note that in almost any system, an "admin" effectively has access to all data. In something like oC10, as an example, an admin can create a new user with a password that the admin knows, add that user to a group, then login as that new "dummy" user and see data that the group has access to. It gets a bit harder when the installation has an IDP and users, groups, group membership all comes from some LDAP source. Then the admin is more restricted - if there are no local users or groups allowed, and the admin cannot add themselves to any group, then they might find it difficult to get access to data. Similar will happen in oCIS if all the user and group data comes from the IDP/LDAP.

There could be a system setting/config/role where the local admin account is not allowed to be a member of any group, not allowed to have any data access to spaces, is not allowed to receive shares... The local admin account would not be a "real user" for file-sync-and-share.

If there was some way to get space roles/permissions from the IDP/LDAP (and for the admins of the LDAP to define those) then an installation could be set up so that the local admin account did not need to be used at all for space access management.

Just a few thoughts for future designs.

**Comment by dragonchaser** [comment]
Limitation on the side of the graph api is very expensive, I suggest to passt this to web. I would just handle the 404 graceful. Also this breaks the usercount in the top right of the spaces overview.

![image](https://user-images.githubusercontent.com/1058116/201036659-85acacf4-da09-44da-ac90-bd6dc8d37661.png)

**Comment by dragonchaser** [comment]
Also @wkloucek  pointed out, when rendering the space overview every user is requested to create a user count. IMHO this does not make sense, displaying the size of the permissions object should be sufficient.

**Comment by mmattel** [comment]
> I will just note that in almost any system, an "admin" effectively has access to all data.

Warning, we mix up two different things. While technically the above sentence is fully correct, there is also another aspect to be considered which is legal and tracking. We should not underestimate the latter. If an admin is by design always possible to look into content, any access can never be distinguished if in scope or not. If there is by design a differenciation, he can access data as before including attempts to fake stuff, but compared to before things get logged and are traceable which is for security and legal a big step forward.

My point is, that we should not underestimate the **benefit** of having this seperation - from a security, legal and marketing pov. Just see how crazy the wold today is, having such a possibility is a clear selling point and differenciator.

**Comment by phil-davis** [comment]
@mmattel yes, I agree. authentication settings changes (adding users, group membership, adding.adjusting permissions...) need to all be logged in a clear way, etc. That way inappropriate admin actions can be traced.
